### PR TITLE
MVP for agenda api

### PR DESCRIPTION
**Pull request by AlexandreJunod**
### Required points before merging embed-agenda in develop
- [x] Secure image
- [x] Filter by domain
- [x] Add agenda-embed build from @monodo
- [x] Restrict fields according to visible submissions, according to domain
- [x] Fix and refactor SITE_DOMAIN and get_absolute_url
- [x] Fix bad domain given
- [x] Return original image size in serializer
- [x] Make filter `startAt` and `endAt` work properly
- [x] `Publication agenda` instead of `Publication calendrier`
- [x] **CANCELED** Dinstinct filters in agenda, bug mutiplies X times. X being the number of forms
- [x] In JS, hide `filter for api` in `fields` when type is not correct
- [x] Make poster mandatory in fixture
- [x] Check that featured item comes first
- [x] Remove agenda app that was there for developping purposes only
- [x] ⚠️ **WRITE TESTS**

@monodo 
- [x] Check agenda with every submission status (To be sure we defined it correctly)
- [x]  Check if order of items in agenda is correct, or should it be changed on QS 🐛 ? featured items do NOT come first
- [x] Check if `featured` (checkbox in admin) work properly 🐛 ? featured items do NOT come first
- [x] Check if `filters` in `/rest/agenda` are isolated correctly with large data (preprod)
- [x] Check if `filters` in `/rest/agenda/domain` return only :
   - [x] current domain (entity)
- [x] Check security of images `/rest/image/` and `/rest/image/thumbor/`
- [x] Make a last review
- [x] Review documentation https://github.com/yverdon/geocity/wiki/agenda 
- [x] Resolve all comments

@AlexandreJunod 
- [x] Check agenda with every submission status (To be sure we defined it correctly)
- [x] Check if order of items in agenda is correct, or should it be changed on QS
- [x] Check if `featured` (checkbox in admin) work properly
   - [x] Check if the "bad code" is removed for `featured` in serializer
- [x] Check if `filters` in `/rest/agenda` are isolated correctly with large data (preprod)
- [x] Check if `filters` in `/rest/agenda` return only :
   - [x] current domain (entity)
- [ ] Check security of images `/rest/image/` and `/rest/image/thumbor/`
- [x] Read all the code and remove every `TODO`. Tag `IMPROVE V2` things not required in V1
- [x] Make a last review
- [x] Resolve all comments

### CODE REVIEW

File: geocity/apps/agenda/views.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import os
+
+from django.conf import settings
+from django.http import FileResponse, JsonResponse
+from django.shortcuts import render
+
+
+# TODO: Place this view in the form or submission, so we have direct access to the image settings, public or not ? From API
```
Comment (AlexandreJunod) [inline geocity/apps/agenda/views.py:8]: Remove

File: geocity/apps/api/serializers.py [hunk 2]
```diff
@@ -756,3 +753,207 @@ class SubmissionFiltersSerializer(serializers.Serializer):
...
+        if aggregated_geotime_qs["submission_geo_time_end_date"]
+        else ""
+    )
+
+    # Rewrite poster to match agenda-embed
+    if "poster" in result["properties"]:
+        result["properties"]["poster"] = {
+            "src": f'http://localhost:9095/agenda/image/display/{result["properties"]["poster"]}',
+            "width": "1365",
+            "height": "2048",
```
Comment (AlexandreJunod) [inline geocity/apps/api/serializers.py:900]: Define how we can have width and height ? Is it the actual image width and height ?

File: geocity/apps/api/serializers.py [hunk 3]
```diff
@@ -756,3 +753,207 @@ class SubmissionFiltersSerializer(serializers.Serializer):
...
+        }
+    # print(result["properties"]["poster"])
```
Comment (AlexandreJunod) [inline geocity/apps/api/serializers.py:902]: Delete

File: geocity/apps/api/views.py [hunk 4]
```diff
@@ -466,3 +468,91 @@ def get_queryset(self):
...
+
+    throttle_scope = "agenda"
+    serializer_class = serializers.AgendaSerializer
+    pagination_class = AgendaResultsSetPagination
+    permission_classes = [permissions.AllowAllRequesters]
+
+    def get_queryset(self):
+        """
+        This view should return a list of events for agenda with minimal information when id is not given.
+        Use used_as_api_filter to filter field by values, values (id = line of the value)
```
Comment (AlexandreJunod) [inline geocity/apps/api/views.py:492]: Use used_as_api_filter to filter field by values. Each line of  MultipleChoiceField or SingleChoiceField is considered as an "id"

File: geocity/apps/submissions/management/commands/fixturize.py [hunk 5]
```diff
@@ -110,6 +111,18 @@ def handle(self, *args, **options):
...
+        )
+
+        if os.path.exists(image_dir):
+            shutil.rmtree(image_dir)
+
+        source_dir = (
+            "geocity/apps/submissions/management/fixturize_data/images/posters/"
+        )
+        shutil.copytree(source_dir, image_dir)
+
```
Comment (AlexandreJunod) [inline geocity/apps/submissions/management/commands/fixturize.py:125]: Improve this and put in a function

File: geocity/tests/factories.py [hunk 6]
```diff
@@ -464,7 +464,7 @@ class Meta:
     placeholder = factory.Faker("word")
     help_text = factory.Faker("word")
     input_type = forms_models.Field.INPUT_TYPE_TEXT
-    is_public_when_permitrequest_is_public = True
+    public_info = True
```
Comment (monodo) [inline geocity/tests/factories.py:467]: refactor one name is fixed

File: geocity/tests/submissions/functional/test_submission_api.py [hunk 7]
```diff
@@ -140,7 +140,7 @@ def setUp(self):
             submission=self.draft_submission_1,
             form=self.forms[0],
         )
-        field = factories.FieldFactory(is_public_when_permitrequest_is_public=True)
+        field = factories.FieldFactory(public_info=True)
```
Comment (monodo) [inline geocity/tests/submissions/functional/test_submission_api.py:143]: idem

**Review by monodo — CHANGES_REQUESTED** [review]

File: geocity/apps/agenda/views.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+def image_display(request, form_id, image_name):
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/agenda/views.py:8]: Technically no, but if it's the case, we should not put this in an app called "agenda".
We should use another name, so we can group data providers in a specific app for this usage. 
The reason to put this in another app is to make it easier to find all the code that has the same purpose, and to have a complete URL that doesn't start with `/submissions/`

File: geocity/apps/api/views.py [hunk 9]
```diff
@@ -9,6 +9,7 @@
 
 from geocity import geometry
 from geocity.apps.django_wfs3.mixins import WFS3DescribeModelViewSetMixin
+from geocity.apps.django_wfs3.pagination import AgendaResultsSetPagination
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/api/views.py:12]: No, beacause it's only used on agenda route
`api.views.AgendaViewSet` has `pagination_class = AgendaResultsSetPagination`

File: geocity/apps/api/views.py [hunk 10]
```diff
@@ -466,3 +468,93 @@ def get_queryset(self):
...
+        # List every available filter
+        # TODO: Improve queryset to secure it. Actually we can just tag something as used_as_api_filter and it will appear
+        # TODO: Filter this by featured, cause the order matters, agenda-embed has no logic to order elements
+        available_filters = serializers.get_available_filters_for_agenda_as_qs(None)
+
+        # List params given by the request as query_params
+        query_params = self.request.query_params
+
+        # Do the required actions fo every query_param
+        for field_name in query_params:
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/api/views.py:524]: How much should we limit ? Should it be dynamic ? (count of possible params ?)

File: geocity/apps/core/static/js/agenda-embed/agenda-embed.js
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/core/static/js/agenda-embed/agenda-embed.js:1]: Used only for tests, will be removed when ready to merge the PR

File: geocity/apps/forms/models.py [hunk 11]
```diff
@@ -531,6 +531,13 @@ class Form(models.Model):
             Pour une demande anonyme https://geocity.ch/submissions/anonymous/?form=URL_COURTE"""
         ),
     )
+    agenda_visible = models.BooleanField(
+        _("Visible dans l'agenda"),
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/forms/models.py:535]: Added on help text. Not on the field name to don't make it too long, is it good ?

File: geocity/apps/forms/models.py [hunk 12]
```diff
@@ -531,6 +531,13 @@ class Form(models.Model):
...
+        default=False,
+        help_text=_(
+            """Lorsque cette case est cochée, les données de ce formulaire sont accessibles dans l'API <b>/rest/agenda/</b>"""
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/forms/models.py:538]: Done + todo added on the right place on the code

File: geocity/apps/submissions/models.py [hunk 13]
```diff
@@ -1809,6 +1809,15 @@ class SubmissionAmendField(models.Model):
         verbose_name=_("Groupe des administrateurs"),
         limit_choices_to={"permit_department__is_integrator_admin": True},
     )
+    api_light = models.BooleanField(
+        _("Visible dans l'API light"),
+        default=False,
+        help_text=_(
+            """Lorsque cette case est cochée, ce champ est affichée dans la version light de l'api (/rest/RESSOURCE).<br>
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/submissions/models.py:1816]: **"si l'annonce est rendue publique par le pilote"** added on the help text, like the comment above

File: geocity/apps/api/serializers.py (hunk 2 above)
Comment (AlexandreJunod) [reply geocity/apps/api/serializers.py:913]: This is not how it's working. I've been doing some tests, and this value is required to show an image. We can still resize the image through `Thumblor`. Seems like to be the original file dimensions

File: geocity/apps/api/views.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/api/views.py:571]: Done

File: geocity/apps/api/serializers.py [hunk 14]
```diff
@@ -756,3 +753,206 @@ class SubmissionFiltersSerializer(serializers.Serializer):
...
+        "id",
+        "form__name",
+        "form__api_name",
+        "form__category__name",
+        "field_values__field__public_info",
+        "field_values__field__api_light",
+        "field_values__field__used_as_api_filter",
+        "submission__administrative_entity",
+        "submission__author",
+    )
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/api/serializers.py:787]: Done in line 802

File: geocity/apps/api/serializers.py [hunk 15]
```diff
@@ -756,3 +753,206 @@ class SubmissionFiltersSerializer(serializers.Serializer):
...
+    """
+    obj = value.get_selected_forms().all()
+    form_fields = obj.values(
+        "field_values__field__name",
+        "field_values__field__api_name",
+        "field_values__field_id",
+        "field_values__field__input_type",
+        "field_values__value__val",
+        "form_id",
+        "form__amend_fields__amend_field_value__value",
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/api/serializers.py:775]: form__amend_fields__amend_field_value__value used to know if something is "featured" (mis en avant)

File: geocity/apps/api/views.py [hunk 16]
```diff
@@ -466,3 +476,162 @@ def get_queryset(self):
...
+def image_thumbor_display(request, submission_id, image_name):
+    width = request.GET.get("width", 397)
+    height = request.GET.get("height", 562)
+    format = request.GET.get("format", "webp")
+    fit = request.GET.get("fit", None)
+
+    INTERNAL_WEB_ROOT_URL = "http://web:9000"
+    image_url = f"{INTERNAL_WEB_ROOT_URL}/rest/image/{submission_id}/{image_name}"
+
+    # TODO: understand (adaptive-)(full-)fit-in between unsafe and size
```
Comment (monodo) [inline geocity/apps/api/views.py:517]: optional => v2

File: geocity/apps/forms/models.py [hunk 17]
```diff
@@ -870,7 +877,7 @@ class Field(models.Model):
         max_length=255,
         blank=True,
         help_text=_(
-            'Ex: "Yverdon-les-Bains" afin de limiter les recherches à Yverdon, <a href="https://api3.geo.admin.ch/services/sdiservices.html#search" target="_blank">Plus d\'informations</a>'
+            'Ex: "Yverdon-les-Bains" afin de limiter les recherches à Yverdon, <a href="https://api3.Sgeo.admin.ch/services/sdiservices.html#search" target="_blank">Plus d\'informations</a>'
```
Comment (monodo) [inline geocity/apps/forms/models.py:880]: typo

File: geocity/apps/agenda/templates/agenda/agenda.html [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+      --agd-headings-font-weight: 500;
+      /* ... */
+    }
+  </style>
+
+  <!-- Render the component -->
+  <agenda-embed api-base-url="http://localhost:9095/rest"></agenda-embed>
+
+  <!-- Initialize the component -->
+  <script type="module" src="{% static 'js/agenda-embed/agenda-embed.js' %}"></script>
```
Comment (monodo) [inline geocity/apps/agenda/templates/agenda/agenda.html:18]: Remove js code from static and use code that is made available throught build process in statics

**Review by monodo — CHANGES_REQUESTED** [review]
+ think about updating the wiki

File: geocity/apps/api/serializers.py [hunk 19]
```diff
@@ -756,3 +753,206 @@ class SubmissionFiltersSerializer(serializers.Serializer):
...
+
+    return result
+
+
+def get_available_filters_for_agenda_as_qs(user):
+    """
+    Returns a list of filters available for a specific user.
+    The order is important, agenda-embed has no logic, everything is set here
+    """
+    # TODO: Improve queryset to secure it. Actually we can just tag something as used_as_api_filter and it will appear
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/api/serializers.py]: Done

File: geocity/tests/factories.py (hunk 6 above)
Comment (AlexandreJunod) [reply geocity/tests/factories.py]: Done

File: geocity/tests/submissions/functional/test_submission_api.py (hunk 7 above)
Comment (AlexandreJunod) [reply geocity/tests/submissions/functional/test_submission_api.py]: Done

File: geocity/apps/forms/models.py (hunk 17 above)
Comment (AlexandreJunod) [reply geocity/apps/forms/models.py]: Done

File: geocity/apps/agenda/templates/agenda/agenda.html (hunk 18 above)
Comment (AlexandreJunod) [reply geocity/apps/agenda/templates/agenda/agenda.html]: Done

File: geocity/apps/agenda/views.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+    image_dir = settings.PRIVATE_MEDIA_ROOT
+
+    image_path = os.path.join(
+        image_dir, f"permit_requests_uploads/{form_id}/{image_name}"
+    )
+
+    # TODO: Ajouter de la sécurité afin de savoir si l'image peut-être affichée ou non
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/agenda/views.py]: Done

**Comment by monodo** [comment]
image

Add to the comment: "Le pilote peut alors contrôler la publication dans l'agenda dans l'onglet traitement

**Comment by monodo** [comment]
Looks like the ordering of items is mixed up:

image

File: geocity/tests/api/test_agenda_api.py [hunk 21]
```diff
@@ -1,18 +1,743 @@
...
+        response = self.client.get(reverse("agenda-list"), {"domain": "sit"})
+        response_json = response.json()
+
+        # Check if request is ok
+        self.assertEqual(response.status_code, 200)
+
+        # Check if filters is not None
+        self.assertNotEqual(response_json["filters"], None)
+
+    # TODO: Count number of elements
```
Comment (monodo) [inline geocity/tests/api/test_agenda_api.py:648]: Resolve TODO ?

File: geocity/apps/django_wfs3/pagination.py [hunk 22]
```diff
@@ -15,3 +17,27 @@ def get_paginated_response(self, data):
                 **data,
             }
         )
+
+
+class AgendaResultsSetPagination(PageNumberPagination):
```
[the remark this answers is not in the document]
Comment (monodo) [reply geocity/apps/django_wfs3/pagination.py:22]: Please move that out of django_wfs3 module

**Review by monodo — CHANGES_REQUESTED** [review]

**Review by monodo — APPROVED** [review]

**Pull request merged**